keras_cv/models/object_detection/faster_rcnn.py

```diff
@@ -226,7 +231,7 @@ def get_config(self):
 
 
 # TODO(tanzheny): add more configurations
-class FasterRCNN(tf.keras.Model):
+class FasterRCNN(ObjectDetectionBaseModel):
```

Comment:
It gives us the in-graph CPU decoding logic in the override of `make_predict_function`.

I also think it's a good step toward standardizing the APIs across OD models to make them all subclass this. wdyt?

Reply:
While it's nice to re-use logic through subclassing and override, it also somewhat defeats the purpose of the repo, which is to provide fork-able model code as a product, and less indirect abstractions.

For example, do we know if all OD models should inherit from this, or just anchor based models that need NMS?

If it's possible, I'd prefer not inherit from base model, and also remove base model from our repo.

Reply:
> If it's possible, I'd prefer not inherit from base model, and also remove base model from our repo.

Good point on the readability, I'm very torn on this.  On one hand, you're correct that the indirection is hard to follow.

On the other hand, OD base model has a LOT of code in it, and a LOT of logic in it; none related to modelling.

Maybe we can upstream keras APIs for asynchronous CPU decoding in the tf.data pipeline, and some sort of post-processing step that runs on CPU before metrics as Keras APIs.

Reply:
I tend to agree that if possible we should avoid having the OD base model.

Right now (until we have NMS decoding working on TPU without the infra of the OD base model), this would result in a lot of duplicate code.

We could consider a refactor to try to use composition instead of inheritance, but that could get ugly since the shared logic is an override of base class methods.

Reply:
+1 for keeping ODBaseModel as an API testing ground for now, then upstreaming common behaviors to Keras core (if possible and if desirable) and then removing ODBaseModel.
Shipping models with tons of duplicating code will not help readability. For people forking the models, in the short term, they will have to fork ODBaseModel too.

Reply:
What duplicate code are exactly required here, if we choose not to inherit?

Reply:
The big thing is `make_predict_function`, but we probably can/should eventually upstream more shared logic (e.g. a default NMS decoder?)

Reply:
IMHO, it's a tradeoff -- we only provide base classes when pros outweighs cons. The prediction / decoding is not a strong selling point 1) compared to asking users to subclass ODBaseModel when they implement YOLO3, so I wonder if we should resist the tempt to do so here, 2) for users that read our code.

At the end of the day, it's an advantage when we can merge common logic which is hard to deal with -- for example, export saved model for TFLite, or to support mixed precision, which is a difficult task and many users don't want to deal with it. But try to avoid another indirect layer seems more reasonable to me.

I will leave this up to you.

Reply:
I think you're right that these should be split in the long-run. For now, I think the `make_predict_function` code is big enough and sensitive enough that it justifies avoiding duplicating the code. 

I'm going to put a TODO for now to eventually scrap the ODBaseModel once we feel like we have a stable API and (hopefully) once the TPU compat issues are solved upstream.